src/general/world_effects/patch_events/MeteorImpactEvent.cs

```diff
@@ -0,0 +1,308 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Linq;
+using Godot;
+using Newtonsoft.Json;
+using Xoshiro.PRNG64;
+
+[JSONDynamicTypeAllowed]
+public class MeteorImpactEvent : IWorldEffect
+{
+    private const string TemplateBiomeForChunks = "patch_event_template_biome";
+
+    [JsonProperty]
+    private readonly HashSet<int> modifiedPatchesIds = new();
+
+    [JsonProperty]
+    private readonly XoShiRo256starstar random;
+
+    [JsonProperty]
+    private GameWorld targetWorld;
+
+    [JsonProperty]
+    private Meteor selectedMeteor = null!;
+
+    public MeteorImpactEvent(GameWorld targetWorld, long randomSeed)
+    {
+        this.targetWorld = targetWorld;
+        random = new XoShiRo256starstar(randomSeed);
+    }
+
+    [JsonConstructor]
+    public MeteorImpactEvent(GameWorld targetWorld, XoShiRo256starstar random)
+    {
+        this.targetWorld = targetWorld;
+        this.random = random;
+    }
+
+    public void OnRegisterToWorld()
+    {
+    }
+
+    public void OnTimePassed(double elapsed, double totalTimePassed)
+    {
+        FinishEvent();
+        TryToTriggerEvent(totalTimePassed);
+    }
+
+    private bool IsSurfacePatch(Patch patch)
+    {
+        return patch.Depth[0] == 0 && patch.BiomeType != BiomeType.Cave;
+    }
+
+    private void TryToTriggerEvent(double totalTimePassed)
+    {
+        if (!AreConditionsMet())
+            return;
+
+        ChooseAffectedPatches();
+        ChooseMeteorType();
+
+        foreach (var patch in targetWorld.Map.Patches.Values)
+        {
+            if (modifiedPatchesIds.Contains(patch.ID))
+            {
+                ChangePatchProperties(patch, totalTimePassed);
+            }
+        }
+
+        LogBeginningOfMeteorStrike();
+    }
+
+    private void ChooseMeteorType()
+    {
+        var meteors = SimulationParameters.Instance.GetAllMeteors().ToList();
+        var index = GetRandomElementByProbability(meteors.Select(meteor => meteor.Probability).ToList(),
+            random.NextDouble());
```

Comment:
This allocates a temporary list. The randomness pick should probably be put into the SimulationParamaters where it can precalculate the list to make the random picking method not allocate memory.